src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiRecordReader.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.servicenow.source;
+
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.plugin.servicenow.source.apiclient.ServiceNowTableAPIClientImpl;
+import io.cdap.plugin.servicenow.source.apiclient.ServiceNowTableDataResponse;
+import io.cdap.plugin.servicenow.source.util.SchemaBuilder;
+import io.cdap.plugin.servicenow.source.util.ServiceNowConstants;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Record reader that reads the entire contents of a ServiceNow table.
+ */
+public class ServiceNowMultiRecordReader extends ServiceNowRecordReader {
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowMultiRecordReader.class);
+  private final ServiceNowMultiSourceConfig multiSourcePluginConf;
+
+  ServiceNowMultiRecordReader(ServiceNowMultiSourceConfig multiSourcePluginConf) {
+    super();
+    this.multiSourcePluginConf = multiSourcePluginConf;
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException {
+    try {
+      if (results == null) {
+        fetchData();
+      }
+
+      if (!iterator.hasNext()) {
+        return false;
+      }
+
+      row = iterator.next();
+
+      pos++;
+    } catch (Exception e) {
+      LOG.error("Error in nextKeyValue", e);
+      throw new IOException("Exception in nextKeyValue", e);
+    }
+    return true;
+  }
+
+  @Override
+  public StructuredRecord getCurrentValue() throws IOException {
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(schema);
+    recordBuilder.set(tableNameField, tableName);
+
+    try {
+      for (Schema.Field field : tableFields) {
+        String fieldName = field.getName();
+        Object fieldValue = convertToValue(fieldName, field.getSchema(), row);
+        recordBuilder.set(fieldName, fieldValue);
+      }
+    } catch (Exception e) {
+      LOG.error("Error decoding row from table " + tableName, e);
+      throw new IOException("Error decoding row from table " + tableName, e);
+    }
+    return recordBuilder.build();
+  }
+
+  public void fetchData() {
+    tableName = split.getTableName();
+    tableNameField = multiSourcePluginConf.getTableNameField();
```

Comment:
Add handling for preview run.

Reply:
The preview run is taking 9:58 minutes for displaying 100 records for tables with approx. 50K records. 
A similar issue was observed in the salesforce-marketing plugin where the following workaround had been proposed to fix the issue. 
https://github.com/data-integrations/salesforce-marketing/pull/16#discussion_r789174283
They are yet to be approved and merged into `develop`.
If we have a general consensus on this fix, then I can go ahead and implement the same for service-now plugins.